### PR TITLE
agent: enable it to be the init process

### LINKED ISSUE #116: Enable agent starting as the init process
Just as the same ability as hyperstart, so that kata-agent can work as the agent of hyperd/runv.

### PR DESCRIPTION
**Pull request by laijs**
Just as the same ability as hyperstart, so that kata-agent
can work as the agent of hyperd/runv.

fixes #116

Signed-off-by: Lai Jiangshan <jiangshanlai@gmail.com>

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -423,6 +480,16 @@ func main() {
 		},
 	}
 
+	if len(os.Args) == 1 && filepath.Base(os.Args[0]) == "init" && os.Getpid() == 1 {
```
Comment (sboeuf) [inline agent.go]: I would prefer if we don't force a specific name here. I think `os.Getpid() == 1` should be enough to know if this agent has been run as the init process.

File: agent.go [hunk 2]
```diff
@@ -380,6 +382,61 @@ func (s *sandbox) teardown() error {
...
 
+type initMount struct {
+	typ, src, des string
+	options       []string
+}
+
+func initAgentAsInit() error {
+	agentLog.Infof("initAgentAsInit(), agent version: %s\n", version)
+
+	mounts := []initMount{
```
Comment (sboeuf) [inline agent.go]: I'd be better to define this outside of this function. This should be const structure list.

File: agent.go [hunk 3]
```diff
@@ -380,6 +382,61 @@ func (s *sandbox) teardown() error {
 	return s.channel.teardown()
 }
 
+type initMount struct {
+	typ, src, des string
+	options       []string
+}
+
+func initAgentAsInit() error {
```
Comment (sboeuf) [inline agent.go]: Please add some verbose comments about the point of this function and explain roughly what it does.

**Comment by laijs** [comment]
@sboeuf updated

File: agent.go [hunk 4]
```diff
@@ -380,6 +381,64 @@ func (s *sandbox) teardown() error {
 	return s.channel.teardown()
 }
 
+type initMount struct {
+	typ, src, des string
```
[the remark this answers is not in the document]
Comment (laijs) [reply agent.go]: updated

**Comment by jshachm** [comment]
@laijs @WeiZhang555 
Do we need to consider the case that using `nsenter` to get into the `agent namespace` and do harm to guest filesystem like `cgroup` or `9p-share dir` ?

**Comment by laijs** [comment]
@jshachm you can build your own rootfs.

**Comment by sboeuf** [comment]
LGTM

**Comment by bergwolf** [comment]
lgtm

**Pull request merged**